chain/client/src/sync/state/mod.rs

```diff
@@ -0,0 +1,388 @@
+        }) as Arc<dyn StateSyncDownloadSource>;
+        let (fallback_source, num_attempts_before_fallback, num_concurrent_requests) =
+            if let SyncConfig::ExternalStorage(ExternalStorageConfig {
+                location,
+                num_concurrent_requests,
+                num_concurrent_requests_during_catchup,
+                external_storage_fallback_threshold,
+            }) = sync_config
+            {
+                let external = match location {
+                    ExternalStorageLocation::S3 { bucket, region, .. } => {
+                        let bucket = create_bucket_readonly(
+                            &bucket,
+                            &region,
+                            timeout.max(Duration::ZERO).unsigned_abs(),
+                        );
+                        if let Err(err) = bucket {
+                            panic!("Failed to create an S3 bucket: {}", err);
+                        }
+                        ExternalConnection::S3 { bucket: Arc::new(bucket.unwrap()) }
+                    }
+                    ExternalStorageLocation::Filesystem { root_dir } => {
+                        ExternalConnection::Filesystem { root_dir: root_dir.clone() }
+                    }
+                    ExternalStorageLocation::GCS { bucket, .. } => ExternalConnection::GCS {
+                        gcs_client: Arc::new(cloud_storage::Client::default()),
+                        reqwest_client: Arc::new(reqwest::Client::default()),
+                        bucket: bucket.clone(),
+                    },
+                };
+                let num_concurrent_requests = if catchup {
+                    *num_concurrent_requests_during_catchup
+                } else {
+                    *num_concurrent_requests
+                } as usize;
+                let fallback_source = Arc::new(StateSyncDownloadSourceExternal {
+                    clock: clock.clone(),
+                    store: store.clone(),
+                    chain_id: chain_id.to_string(),
+                    conn: external,
+                    timeout,
+                }) as Arc<dyn StateSyncDownloadSource>;
+                (
+                    Some(fallback_source),
+                    *external_storage_fallback_threshold as usize,
+                    num_concurrent_requests.min(NUM_CONCURRENT_REQUESTS_FOR_PEERS),
+                )
+            } else {
+                (None, 0, NUM_CONCURRENT_REQUESTS_FOR_PEERS)
+            };
+
+        let (header_validation_sender, header_validation_queue) =
+            tokio::sync::mpsc::unbounded_channel();
+        let (chain_finalization_sender, chain_finalization_queue) =
+            tokio::sync::mpsc::unbounded_channel();
+
+        let downloading_task_tracker = TaskTracker::new(num_concurrent_requests);
+        let downloader = Arc::new(StateSyncDownloader {
+            clock,
+            store: store.clone(),
+            preferred_source: peer_source,
```

Comment:
At the moment we are in an ugly situation with state headers; a node needs to be tracking **all** shards to serve headers for _any_ shard, nodes no longer track all shards, and the strategy for trying to obtain headers from the network is to request them from direct peers of the node at random. It is a remnant of the times when every peer tracked every shard, and it needs to be rewritten entirely. 

Before this PR, if an external source was available we would just directly get the headers from it without any attempts to get it from the network. It looks like we are changing that now and will try the network first for headers. I think it should be OK; just giving a heads up that there is a behavior change hidden here.

Reply:
You mean before the PR, even if we configure to download from peers, we would always fetch the header from external? Hmm. OK.